Overall improvements to bring .mesh export on-par with OpenIV. Further improve skeleton import.

mctrollin opened the pull request:
General: Add vertex color support.
Import: Improve skeleton bone orientations, fix for incorrect secondary bone axis.
Export: Update tangent space support.
Export: Improve bone weights, fix for distorted vertex positions influenced by facial bones.
Export: Adjust unused uv and bone weight format to match OpenIV export.

Commit:
General: Add vertex color support.
Import: Improve skeleton bone orientations, fix for incorrect secondary bone axis.
Export: Update tangent space support.
Export: Improve bone weights, fix for distorted vertex positions influenced by facial bones.
Export: Adjust unused uv and bone weight format to match OpenIV export.

mctrollin commented:
Hi,
I'm currently switching over from 3ds max to blender and was updating this addon during the process. Hope you don't mind.
So testing my changes directly on my assets. But some further testing won't hurt.
The thing with the bone weight adjustment was keeping me awake this week and was also broken in Sollumz. Not sure if GIMS was doing it correctly. Anyways, here we go!
Cheers.

lucasvinbr commented:
Wow, thanks a lot again hahahah! I'll try to test it ASAP

lucasvinbr commented on mesh_geometry_datagather_utils.py:
```diff
@@ -158,25 +158,45 @@ def parse_obj_to_geometrydata(meshObj, parentSkeleton, shaderIndex, correctedNor
-    geom.uvCoords = [(0.0, 0.0)] * len(geom.vertPositions)
-    geom.uvCoords2 = [(0.0, 0.0)] * len(geom.vertPositions)
+    geom.uvCoords = [(1.0, -1.0)] * len(geom.vertPositions)
+    geom.uvCoords2 = [(1.0, -1.0)] * len(geom.vertPositions)
```
hmmmm what's the reasoning for the change in the default value? Could this affect blood textures if the model doesn't have the second UV? Because, in my test, the model got totally red, which wasn't the case before

mctrollin replied:
Technically it should not make a difference. The uvs are just placed on one pixel and this pixel might have a different colour then before. I changed this back for now as it makes probably no difference but if it makes a visual difference it might be worth investigating the best solution. Maybe OpenIV devs decided for 1 -1 for some reason and your model not being red was acutally wrong ^^.

lucasvinbr commented on mesh_geometry_datagather_utils.py:
```diff
@@ -158,25 +158,45 @@ def parse_obj_to_geometrydata(meshObj, parentSkeleton, shaderIndex, correctedNor
+    geom.uvCoords = [(1.0, -1.0)] * len(geom.vertPositions)
+    geom.uvCoords2 = [(1.0, -1.0)] * len(geom.vertPositions)
+    geom.vColor = [(0.0, 0.0, 0.0, 0.0)] * len(geom.vertPositions)
+    geom.vColor2 = [(0.0, 0.0, 0.0, 0.0)] * len(geom.vertPositions)
```
I think we have to replace those tuples by blender's vectors; it's what seems to be used in their place when `loop[uvlayer].uv.copy()` is run. If I try exporting a model with only one vertex color layer, I'm ending up with lots of zeroes in the vColor2 areas of the .mesh file. Like, way more than 4 zeroes hahaha... So, replacing `[(0.0, 0.0, 0.0, 0.0)]` by `[Vector((0.0, 0.0, 0.0, 0.0))]` solved the issue for me

mctrollin replied:
Ah sorry.. that is just me being stupid.

mctrollin replied:
The actual issue is in export_mesh.py where I did just multiply the value 255 times :D.

lucasvinbr replied:
Cool! I just tested it and had to set vColor (1) to be all 1.0, because apparently the ped's shading gets really messed up if it's all zeroes. vColor2 being zero appears to be fine though

lucasvinbr reviewed (changes requested):
So, I've imported and exported a mesh and most stuff look fine, but:

- OpenIV failed to read my exported model (which didn't have a 2nd vertex color layer) unless I changed the vColor2 default values to Vector, as in the comment below
- After the fix, I exported it back from openIV. The weights were messed up because OpenIV seems to export a different .skel file when we export the .ydd. Exporting the .yft returned the correct skeleton, and, after replacing, the weights were fine, Do you get that problem as well?

By the way, I'm using blender 3.6

Commit:
Fix vertex color serialization.
Fix exception if model has no vertex groups.
Fix exception if model has no vertex colors.
Revert default uv coordinates to 0 0 for now.

mctrollin committed:
Merge branch 'lucasvinbr:master' into master

mctrollin commented:
Added a submit to fix issues as you described if the model has no valid channels for weights or vertex colors.

lucasvinbr committed:
Set default vColor1 to be all 255

Ped's shading gets all messed up if it's all zeroes

lucasvinbr commented:
So, I exported some more models and I think everything is fine now. Thanks again!

mctrollin commented:
Good point about the defautl vVolor!
